File: backend/internal/routes/content.go
package routes

import (
	"onflix/internal/controllers"
	"onflix/internal/middleware"
	"onflix/internal/services"

	"github.com/gin-gonic/gin"
)

func SetupPublicContentRoutes(rg *gin.RouterGroup, services *services.Services) {
	contentController := controllers.NewContentController(services)
	authMiddleware := middleware.NewAuthMiddleware(services.DB, services.Config.JWT.Secret)

	// Public content routes (for browsing without subscription)
	content := rg.Group("/content")
	content.Use(authMiddleware.OptionalAuth())
	{
		// Browse content
		content.GET("", contentController.BrowseContent)
		content.GET("/featured", contentController.GetFeaturedContent)
		content.GET("/trending", contentController.GetTrendingContent)
		content.GET("/new-releases", contentController.GetNewReleases)
		content.GET("/originals", contentController.GetOriginals)

		// Content details
		content.GET("/:contentID", contentController.GetContentDetails)
		content.GET("/:contentID/similar", contentController.GetSimilarContent)
		content.GET("/:contentID/trailers", contentController.GetTrailers)

		// Search
		content.GET("/search", contentController.SearchContent)
		content.GET("/search/suggestions", contentController.GetSearchSuggestions)

		// Genres and categories
		content.GET("/genres", contentController.GetGenres)
		content.GET("/genres/:genre", contentController.GetContentByGenre)
		content.GET("/categories", contentController.GetCategories)
		content.GET("/categories/:category", contentController.GetContentByCategory)

		// TV Show specific
		tvshows := content.Group("/tv-shows")
		{
			tvshows.GET("/:showID/seasons", contentController.GetSeasons)
			tvshows.GET("/:showID/seasons/:seasonNumber", contentController.GetSeason)
			tvshows.GET("/:showID/seasons/:seasonNumber/episodes", contentController.GetEpisodes)
			tvshows.GET("/:showID/seasons/:seasonNumber/episodes/:episodeNumber", contentController.GetEpisode)
		}
	}
}

func SetupContentRoutes(rg *gin.RouterGroup, services *services.Services) {
	contentController := controllers.NewContentController(services)
	authMiddleware := middleware.NewAuthMiddleware(services.DB, services.Config.JWT.Secret)

	// Protected content routes (require subscription)
	content := rg.Group("/content")
	content.Use(authMiddleware.RequireSubscription())
	{
		// Video streaming
		content.GET("/:contentID/stream", contentController.StreamVideo)
		content.GET("/:contentID/stream/:quality", contentController.StreamVideoQuality)
		content.POST("/:contentID/stream/token", contentController.GetStreamingToken)

		// TV Show streaming
		content.GET("/tv-shows/:showID/seasons/:seasonNumber/episodes/:episodeNumber/stream", contentController.StreamEpisode)

		// Download for offline viewing
		content.POST("/:contentID/download", contentController.DownloadContent)
		content.GET("/downloads", contentController.GetDownloads)
		content.DELETE("/downloads/:downloadID", contentController.RemoveDownload)

		// Subtitles
		content.GET("/:contentID/subtitles", contentController.GetSubtitles)
		content.GET("/:contentID/subtitles/:language", contentController.GetSubtitleFile)

		// User interactions
		interactions := content.Group("/:contentID")
		{
			interactions.POST("/like", contentController.LikeContent)
			interactions.DELETE("/like", contentController.UnlikeContent)
			interactions.POST("/rate", contentController.RateContent)
			interactions.POST("/review", contentController.AddReview)
			interactions.GET("/reviews", contentController.GetReviews)
		}

		// Continue watching
		content.GET("/continue-watching", contentController.GetContinueWatching)
		content.POST("/:contentID/watch-progress", contentController.UpdateWatchProgress)
	}

	// Recommendations (require auth but not necessarily subscription)
	recommendations := rg.Group("/recommendations")
	{
		recommendations.GET("", contentController.GetRecommendations)
		recommendations.GET("/trending", contentController.GetTrendingRecommendations)
		recommendations.GET("/because-you-watched/:contentID", contentController.GetBecauseYouWatchedRecommendations)
		recommendations.POST("/feedback", contentController.SubmitRecommendationFeedback)
	}
}

File: frontend/src/lib/api/content.ts
import { ApiResponse } from '@/types/api';
import { Content, ContentFilters, SearchParams, Season } from '@/types/content';
import apiClient from './client';
import { Download, DownloadResponse, StreamingResponse, StreamingToken, Subtitle, WatchProgress } from './streaming';



class ContentAPI {
  // Public content endpoints
  async browseContent(filters?: ContentFilters): Promise<ApiResponse<Content[]>> {
    const params = new URLSearchParams();
    if (filters) {
      Object.entries(filters).forEach(([key, value]) => {
        if (value !== undefined) params.append(key, value.toString());
      });
    }
    return apiClient.get(`/api/v1/content?${params.toString()}`);
  }

  async getContentById(id: string): Promise<ApiResponse<Content>> {
    return apiClient.get(`/api/v1/content/${id}`);
  }

  async getFeaturedContent(): Promise<ApiResponse<Content[]>> {
    return apiClient.get('/api/v1/content/featured');
  }

  async getTrendingContent(): Promise<ApiResponse<Content[]>> {
    return apiClient.get('/api/v1/content/trending');
  }

  async getNewReleases(): Promise<ApiResponse<Content[]>> {
    return apiClient.get('/api/v1/content/new-releases');
  }

  async searchContent(params: SearchParams): Promise<ApiResponse<Content[]>> {
    const searchParams = new URLSearchParams();
    Object.entries(params).forEach(([key, value]) => {
      if (value !== undefined) searchParams.append(key, value.toString());
    });
    return apiClient.get(`/api/v1/content/search?${searchParams.toString()}`);
  }

  async getContentByGenre(genre: string, page = 1, limit = 20): Promise<ApiResponse<Content[]>> {
    return apiClient.get(`/api/v1/content/genres/${genre}?page=${page}&limit=${limit}`);
  }

  async getTVShowSeasons(showId: string): Promise<ApiResponse<Season[]>> {
    return apiClient.get(`/api/v1/content/tv-shows/${showId}/seasons`);
  }

  // Protected content endpoints (require authentication)
  async streamContent(contentId: string): Promise<ApiResponse<StreamingResponse>> {
    return apiClient.get(`/api/v1/content/${contentId}/stream`);
  }

  async streamContentWithQuality(contentId: string, quality: string): Promise<ApiResponse<StreamingResponse>> {
    return apiClient.get(`/api/v1/content/${contentId}/stream/${quality}`);
  }

  async getStreamingToken(contentId: string): Promise<ApiResponse<StreamingToken>> {
    return apiClient.post(`/api/v1/content/${contentId}/stream/token`);
  }

  async streamTVShowEpisode(
    showId: string,
    seasonNumber: number,
    episodeNumber: number
  ): Promise<ApiResponse<StreamingResponse>> {
    return apiClient.get(`/api/v1/content/tv-shows/${showId}/seasons/${seasonNumber}/episodes/${episodeNumber}/stream`);
  }

  async downloadContent(contentId: string, quality: string): Promise<ApiResponse<DownloadResponse>> {
    return apiClient.post(`/api/v1/content/${contentId}/download`, { quality });
  }

  async getUserDownloads(): Promise<ApiResponse<Download[]>> {
    return apiClient.get('/api/v1/content/downloads');
  }

  async removeDownload(downloadId: string): Promise<ApiResponse> {
    return apiClient.delete(`/api/v1/content/downloads/${downloadId}`);
  }

  async getSubtitles(contentId: string): Promise<ApiResponse<Subtitle[]>> {
    return apiClient.get(`/api/v1/content/${contentId}/subtitles`);
  }

  async getSubtitleFile(contentId: string, language: string): Promise<string> {
    const response = await apiClient.get(`/api/v1/content/${contentId}/subtitles/${language}`);
    return response.data as string;
  }

  async rateContent(contentId: string, rating: RatingRequest): Promise<ApiResponse> {
    return apiClient.post(`/api/v1/content/${contentId}/rate`, rating);
  }

  async addReview(contentId: string, review: ReviewRequest): Promise<ApiResponse> {
    return apiClient.post(`/api/v1/content/${contentId}/review`, review);
  }

  async updateWatchProgress(contentId: string, progress: WatchProgress): Promise<ApiResponse> {
    return apiClient.post(`/api/v1/content/${contentId}/watch-progress`, progress);
  }
}

const contentAPI = new ContentAPI();
export default contentAPI;

